packages/crypto/src/ed25519.rs

```diff
@@ -48,23 +55,127 @@ pub fn ed25519_verify(message: &[u8], signature: &[u8], public_key: &[u8]) -> Cr
+    let messages_len = messages.len();
+    let signatures_len = signatures.len();
+    let public_keys_len = public_keys.len();
+
+    let mut messages = messages.to_vec();
+    let mut public_keys = public_keys.to_vec();
+    if messages_len == signatures_len && messages_len == public_keys_len {
+        if messages_len == 0 {
+            // Verifying nothing equals false
+            return Ok(false);
+        }
+    } else if messages_len == 1 && signatures_len == public_keys_len {
+        if signatures_len == 0 {
+            return Err(CryptoError::batch_err(
+                "No signatures / public keys provided",
+            ));
+        }
+        // Replicate message, for multisig
+        messages = messages.repeat(signatures_len);
+    } else if public_keys_len == 1 && messages_len == signatures_len {
+        if messages_len == 0 {
+            return Err(CryptoError::batch_err("No messages / signatures provided"));
+        }
+        // Replicate pubkey
+        public_keys = public_keys.repeat(messages_len);
+    } else {
+        return Err(CryptoError::batch_err(
+            "Mismatched / erroneous number of messages / signatures / public keys",
+        ));
+    }
+    debug_assert_eq!(messages.len(), signatures_len);
+    debug_assert_eq!(messages.len(), public_keys.len());
+
+    let mut batch = ed25519::batch::Verifier::new();
+
+    for (((i, &message), &signature), &public_key) in
+        (1..).zip(&messages).zip(signatures).zip(&public_keys)
+    {
+        // Validation
+        if message.len() > MESSAGE_MAX_LEN {
+            return Err(CryptoError::msg_err(format!(
+                "message {}: too large: {}",
+                i,
+                message.len()
+            )));
+        }
+        if signature.len() != EDDSA_SIGNATURE_LEN {
+            return Err(CryptoError::sig_err(format!(
+                "signature {}: wrong / unsupported length: {}",
+                i,
+                signature.len()
+            )));
+        }
+        let pubkey_len = public_key.len();
+        if pubkey_len == 0 {
+            return Err(CryptoError::pubkey_err(format!("public key {}: empty", i)));
+        }
+        if pubkey_len != EDDSA_PUBKEY_LEN {
+            return Err(CryptoError::pubkey_err(format!(
+                "public key {}: wrong / unsupported length: {}",
+                i, pubkey_len,
+            )));
+        }
+
+        // Deserialization
+        let signature = ed25519::Signature::try_from(signature).map_err(|err| {
+            CryptoError::generic_err(format!("signature {}: {}", i, err.to_string()))
+        })?;
+
+        let public_key = ed25519::VerificationKey::try_from(public_key).map_err(|err| {
+            CryptoError::generic_err(format!("public key {}: {}", i, err.to_string()))
+        })?;
+
+        // Enqueing
+        batch.queue((public_key.into(), signature, message));
+    }
+
+    // Batch verification
+    match batch.verify(&mut OsRng) {
+        Ok(_) => Ok(true),
```

Comment:
Can we use

```suggestion
        Ok(()) => Ok(true),
```

to ensure no data is lost here?